ts/editor-toolbar/index.ts

```diff
@@ -0,0 +1,153 @@
+    hidden?: boolean;
+}
+
+function searchByIdOrIndex<T extends { id?: string }>(
+    values: T[],
+    idOrIndex: string | number
+): T | undefined {
+    return typeof idOrIndex === "string"
+        ? values.find((value) => value.id === idOrIndex)
+        : values[idOrIndex];
+}
+
+function showComponent(component: Hideable) {
+    component.hidden = false;
+}
+
+function hideComponent(component: Hideable) {
+    component.hidden = true;
+}
+
+function toggleComponent(component: Hideable) {
+    component.hidden = !component.hidden;
+}
+
+class EditorToolbar extends HTMLElement {
+    component?: SvelteComponent;
+
+    buttons = writable(defaultButtons);
+    menus = writable(defaultMenus);
+
+    connectedCallback(): void {
+        setupI18n({ modules: [ModuleName.EDITING] }).then(() => {
+            this.component = new EditorToolbarSvelte({
+                target: this,
+                props: {
+                    buttons: this.buttons,
+                    menus: this.menus,
+                    nightMode: checkNightMode(),
+                },
+            });
+        });
+    }
+
+    updateButtonGroup<T>(
+        update: (component: DynamicSvelteComponent<typeof ButtonGroup> & T) => void,
+        group: string | number
+    ): void {
+        this.buttons.update(
+            (
+                buttonGroups: (DynamicSvelteComponent<typeof ButtonGroup> &
+                    ButtonGroupProps)[]
+            ) => {
+                const foundGroup = searchByIdOrIndex(buttonGroups, group);
+
+                if (foundGroup) {
+                    update(
+                        foundGroup as DynamicSvelteComponent<typeof ButtonGroup> &
+                            ButtonGroupProps &
+                            T
+                    );
+                }
+
+                return buttonGroups;
+            }
+        );
+    }
+
+    showButtonGroup(group: string | number): void {
+        this.updateButtonGroup<Hideable>(showComponent, group);
+    }
+
+    hideButtonGroup(group: string | number): void {
+        this.updateButtonGroup<Hideable>(hideComponent, group);
+    }
+
+    toggleButtonGroup(group: string | number): void {
+        this.updateButtonGroup<Hideable>(toggleComponent, group);
+    }
+
+    updateButton<T>(
+        update: (component: DynamicSvelteComponent & T) => void,
+        group: string | number,
+        button: string | number
+    ): void {
+        this.updateButtonGroup(
+            (
+                foundGroup: DynamicSvelteComponent<typeof ButtonGroup> &
+                    ButtonGroupProps
+            ) => {
+                const foundButton = searchByIdOrIndex(
+                    foundGroup.buttons as (DynamicSvelteComponent & { id?: string })[],
+                    button
+                );
+
+                if (foundButton) {
+                    update(foundButton as DynamicSvelteComponent & T);
+                }
+            },
+            group
+        );
+    }
```

Comment:
It feels like there's a lot of generics and casting going on here, but that may just be my lack of TypeScript experience talking? The way I expected it to work is to have some interface that every entry in the store conforms to, like the following (which may not be valid typescript)

```typescript
interface ToolbarItem extends DynamicSvelteComponent<any> {
  id?: string,
  hidden: boolean,
  active: boolean,
}
```

Would that make sense?